### PR TITLE
Initial version of the module.

### PR DESCRIPTION
**Pull request by shackijj**
Subj.

### CODE REVIEW

File: src/Gridmap/utils/hexagonGrid.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                [cos(240) + x, sin(240) + y],
+                [cos(300) + x, sin(300) + y],
+                [cos(0) + x, sin(0) + y]
+            ];
+
+            const hexagonGlobals = hexagon
+                .map(([x, y]) =>
+                    projection.fromGlobalPixels(
+                        [
+                            offsetLeft + (x * R),
```
Comment (ruslankhh) [inline src/Gridmap/utils/hexagonGrid.js]: What happens if the developer uses a different format for points (latlong)?
This also applies to squares.

File: src/Gridmap/Gridmap.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    function getRequiredOption(options, path) {
+        const value = get(options, path);
+        if (!value) {
+            throw new Error(`options.${path} is required parameter`);
+        }
+        return value;
+    }
+    class Gridmap {
+        constructor(options) {
+            const map = getRequiredOption(options, 'map');
```
[the remark this answers is not in the document]
Comment (shackijj) [reply src/Gridmap/Gridmap.js]: This module can't work without map. I could implement that but it won't change behaviour of the module and add complexity in the code.

What benefits do you see by doing the changes you mentioned?

File: src/Gridmap/utils/hexagonGrid.js (hunk 1 above)
Comment (shackijj) [reply src/Gridmap/utils/hexagonGrid.js]: Nothing. We are woking with pixel coordinates here not geographical.

File: src/Gridmap/utils/hexagonGrid.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+
+export function sin(angle) {
+    return Math.sin(Math.PI * angle / 180);
+}
+
+export function cos(angle) {
+    return Math.cos(Math.PI * angle / 180);
+}
+
+export default function hexagonGrid(projection, zoom, R, offsetLeft, offsetTop, width, height) {
```
[the remark this answers is not in the document]
Comment (shackijj) [reply src/Gridmap/utils/hexagonGrid.js]: It's a private function I don't think that changing the code like in the example below will really improve it.
```
export default function hexagonGrid({projection, zoom, R, offsetLeft, offsetTop, width, height}) {
...

hexagonGrid({
    projection,
    zoom: 100,
});
```

If you insist I could change that, but there aren't any tests which will confirm that my changes won't the break the logic.

File: src/Gridmap/utils/hexagonGrid.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    const colWidth = 1.5 * R;
+    const rowHeight = 1.5 * R;
+    const cols = Math.floor((width + (R / 2)) / colWidth) + 1;
+    const rows = Math.floor(height / rowHeight);
+
+    const result = {type: 'FeatureCollection', features: []};
+
+    let id = 0;
+    for (let c = 0; c < cols; c++) {
```
[the remark this answers is not in the document]
Comment (shackijj) [reply src/Gridmap/utils/hexagonGrid.js]: What kind of comments would you like to see here?
The code is pretty straightforward. There is no 'if' or switch sentences.
I think that providing tests for the code will be more valuable, but it worth doing outside the scope of this review. I created the issues about tests #4 and will proceed with it as soon as we decide that will use this implementation for the project (not the canvas rendering).

File: src/Gridmap/utils/hexagonGrid.js (hunk 1 above)
Comment (ruslankhh) [reply src/Gridmap/utils/hexagonGrid.js]: Okay)

File: src/Gridmap/utils/hexagonGrid.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruslankhh) [reply src/Gridmap/utils/hexagonGrid.js]: Okey.

File: src/Gridmap/utils/hexagonGrid.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    return Math.sin(Math.PI * angle / 180);
+}
+
+export function cos(angle) {
+    return Math.cos(Math.PI * angle / 180);
+}
+
+function hexagonGrid(map, zoom, R, offsetLeft, offsetTop, width, height) {
+    const SIN_OF_SIXTY = sin(60);
+    const colWidth = 1.5 * R;
```
Comment (shuhrat) [inline src/Gridmap/utils/hexagonGrid.js]: `Magic Number` detected =)

File: src/Gridmap/utils/hexagonGrid.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    const result = {type: 'FeatureCollection', features: []};
+    const projection = map.options.get('projection');
+
+    let id = 0;
+    for (let c = 0; c < cols; c++) {
+        for (let r = 0; r < rows; r++) {
+            const horizontalShift = (c % 2 === 0) ? 0 : -1 * SIN_OF_SIXTY;
+            const x = c * 1.5;
+            const y = r * (2 * SIN_OF_SIXTY) + horizontalShift;
+            const hexagon = [
```
Comment (shuhrat) [inline src/Gridmap/utils/hexagonGrid.js]: может вынести в функцию?

File: src/Gridmap/utils/hexagonGrid.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                .map((point) => {
+                    return projection.fromGlobalPixels([offsetLeft + (point[0] * R), offsetTop + (point[1] * R)], zoom);
+                })
+                .map((coords) => {
+                    return [coords[1], coords[0]];
+                });
+
+            result.features.push({
+                type: 'Feature',
+                id: 'hxg' + id++,
```
Comment (shuhrat) [inline src/Gridmap/utils/hexagonGrid.js]: string templates? `hxg${id++}`

хотя id и есть c+r

File: src/Gridmap/utils/hexagonGrid.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                [cos(60) + x, sin(60) + y],
+                [cos(120) + x, sin(120) + y],
+                [cos(180) + x, sin(180) + y],
+                [cos(240) + x, sin(240) + y],
+                [cos(300) + x, sin(300) + y],
+                [cos(0) + x, sin(0) + y]
+            ];
+            const hexagonGlobals = hexagon
+                .map((point) => {
+                    return projection.fromGlobalPixels([offsetLeft + (point[0] * R), offsetTop + (point[1] * R)], zoom);
```
Comment (shuhrat) [inline src/Gridmap/utils/hexagonGrid.js]: у нас и так хромает производительность, так почему плодить массивы данных
```js
const coords = projection.fromGlobalPixels(...)
return [coords[1], coords[0]];
```

File: src/Gridmap/Gridmap.js [hunk 9]
```diff
@@ -12,6 +12,35 @@ ymaps.modules.define('Gridmap', ['Polygonmap', 'util.bounds'], (provide, Polygon
         return value;
     }
     class Gridmap {
+        /**
+         * @typedef {Object} GridBounds
```
Comment (shuhrat) [inline src/Gridmap/Gridmap.js]: 😍

File: README.md [hunk 10]
```diff
@@ -26,3 +26,83 @@ For testing:
 ```bash
 npm test
 ```
+
+
+## Classes
```
Comment (shuhrat) [inline README.md]: сгенерировал автоматически?

**Pull request merged**